cli/tests/test_config_command.rs

```diff
@@ -602,6 +602,96 @@ fn test_config_set_nontable_parent() {
     "###);
 }
 
+#[test]
+fn test_config_unset_non_existent_key() {
+    let mut test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["git", "init", "repo"]);
+    // Point to a config file since `config set` can't handle directories.
+    let user_config_path = test_env.config_path().join("config.toml");
+    test_env.set_config_path(user_config_path);
+    let repo_path = test_env.env_root().join("repo");
+
+    let stderr = test_env.jj_cmd_failure(&repo_path, &["config", "unset", "--user", "nonexistent"]);
+    insta::assert_snapshot!(stderr, @r###"
+    Config error: configuration property "nonexistent" not found
+    For help, see https://martinvonz.github.io/jj/latest/config/.
+    "###);
+}
+
+#[test]
+fn test_config_unset_for_user() {
+    let mut test_env = TestEnvironment::default();
+    test_env.jj_cmd_ok(test_env.env_root(), &["git", "init", "repo"]);
+    // Point to a config file since `config set` can't handle directories.
+    let user_config_path = test_env.config_path().join("config.toml");
+    test_env.set_config_path(user_config_path.clone());
+    let repo_path = test_env.env_root().join("repo");
+
+    test_env.jj_cmd_ok(&repo_path, &["config", "set", "--user", "foo", "true"]);
+    test_env.jj_cmd_ok(&repo_path, &["config", "unset", "--user", "foo"]);
+
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &["config", "set", "--user", "table.foo", "true"],
+    );
+    test_env.jj_cmd_ok(&repo_path, &["config", "unset", "--user", "table.foo"]);
+
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &["config", "set", "--user", "inline-table", "{ foo = true }"],
+    );
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &["config", "unset", "--user", "inline-table.foo"],
+    );
+
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &[
+            "config",
+            "set",
+            "--user",
+            "table.inline-table",
+            "{ foo = true }",
+        ],
+    );
+    test_env.jj_cmd_ok(
+        &repo_path,
+        &["config", "unset", "--user", "table.inline-table.foo"],
+    );
+
+    let user_config_toml = std::fs::read_to_string(&user_config_path)
+        .unwrap_or_else(|_| panic!("Failed to read file {}", user_config_path.display()));
```

Comment:
nit: just do `.unwrap()` in tests unless there's a special consideration

Reply:
I just copy/pasted that from another test in the same file and haven't really thought about it.
Now that I put some thought into it, I guess the "custom" panic here serves a better error message than a plain `.unwrap()`, so I am inclined to leave it as is.

`.unwrap_or_else(...)`
```
---- test_config_command::test_config_unset_for_user stdout ----
thread 'test_config_command::test_config_unset_for_user' panicked at cli/tests/test_config_command.rs:665:29:
Failed to read file /private/var/folders/hb/zr1tp04d29b8wdh14lczb3c40000gn/T/jj-test-IkBUbD/config/config.toml
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

vs.

`.unwrap()`
```
---- test_config_command::test_config_unset_for_user stdout ----
thread 'test_config_command::test_config_unset_for_user' panicked at cli/tests/test_config_command.rs:664:64:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
---
Another option would be to use `.expect()`
```suggestion
        .expect(&format!("Failed to read file {}", user_config_path.display()));
```
I don't know if that is "better", because of `&format!(...)`. What's your take on this?

Reply:
Well, if `std::fs::read_to_string(&user_config_path)` is likely to fail, it makes sense to provide a better error message. However, we're testing "config unset" (which isn't the command to create new config file), so I would just use `.unwrap()`.

Reply:
Thanks for elaborating. That makes a lot of sense now, since I copied that part from a `jj config set` test case. 💡 
I will replace it with `.unwrap()` then.

Reply:
Do you test unsetting it twice in a row? What would be the expected result for the second time?

Reply:
> Do you test unsetting it twice in a row? What would be the expected result for the second time?

Right now there is no test case, which covers running `jj config unset` twice in a row for the same key.

With the current implementation, given this user config:
```toml
[table]
foo = "bar"
```
..running `jj config --user unset table.foo` twice in a row will result in the following error:
```
Config error: configuration property "table.foo" not found
For help, see https://martinvonz.github.io/jj/latest/config/.
```
..because the first time running `jj config --user unset table.foo` removes `table.foo`. So the second execution can't find the key any longer.

(The same applies for `jj config --repo unset table.foo`)

---

@joyously, your question got me thinking. I added a comment in the issue: https://github.com/martinvonz/jj/issues/4458#issuecomment-2438613512

Reply:
I don't know whether it should be an error or a warning, but it's nice that `jj config unset` reports unknown key to unset (i.e. noop change).